### PR TITLE
feat: add key races action schema

### PR DESCRIPTION
**Pull request by olavoparno**
closes   https://github.com/Stand-With-Crypto/swc-internal/issues/137

fixes 

## How has it been tested?

- [x] Locally
- [x] Vercel Preview Branch
- [ ] Unit test
- [ ] Functional test

### CODE REVIEW

**Review by mhamendes — DISMISSED** [review]

File: src/actions/actionCreateUserActionViewKeyRaces.ts [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+  analytics: ReturnType<typeof getServerAnalytics>
+  peopleAnalytics: ReturnType<typeof getServerPeopleAnalytics>
+}
+
+export const actionCreateUserActionViewKeyRaces = withServerActionMiddleware(
+  'actionCreateUserActionViewKeyRaces',
+  _actionCreateUserActionViewKeyRaces,
+)
+
+async function _actionCreateUserActionViewKeyRaces(input: CreateActionViewKeyRacesInput = {}) {
```
Comment (travisbloom-cb) [inline src/actions/actionCreateUserActionViewKeyRaces.ts:50]: Few product thoughts:
- only persist the action if the user already exists (theyve taken some other action)
- only persist one action of this type per user
   - if the action exists but doesnt have all the district metadata, and it now exists in the new payload, update the existing action
   - if the user has an address and the action doesnt exist or doesnt have the district metadata, use that address as the metadata source of truth

File: src/utils/shared/userActionCampaigns.ts [hunk 2]
```diff
@@ -55,6 +56,9 @@ export enum UserActionVoterAttestationCampaignName {
 export enum UserActionRsvpEventCampaignName {
   DEFAULT = 'DEFAULT',
 }
+export enum UserActionViewKeyRacesCampaignName {
+  DEFAULT = 'DEFAULT',
```
Comment (travisbloom-cb) [inline src/utils/shared/userActionCampaigns.ts:60]: `2024_ELECTION`

File: src/actions/actionCreateUserActionViewKeyRaces.ts (hunk 1 above)
Comment (travisbloom-cb) [reply src/actions/actionCreateUserActionViewKeyRaces.ts:50]: cc @mhamendes

File: src/actions/actionCreateUserActionViewKeyRaces.ts [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+  const analytics = getServerAnalytics({
+    userId: user.id,
+    localUser,
+  })
+  const beforeFinish = () => Promise.all([analytics.flush(), peopleAnalytics.flush()])
+
+  const recentViewKeyRacesUserAction = await getRecentUserActionByUserId(user.id, validatedInput)
+
+  if (recentViewKeyRacesUserAction) {
+    const hasTheSameInfo =
```
Comment (travisbloom-cb) [inline src/actions/actionCreateUserActionViewKeyRaces.ts:91]: this is the logic that needs to change

File: src/actions/actionCreateUserActionViewKeyRaces.ts [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+  waitUntil(beforeFinish())
+  return { user: getClientUser(user) }
+}
+
+async function getRecentUserActionByUserId(
+  userId: User['id'],
+  validatedInput: z.SafeParseSuccess<CreateActionViewKeyRacesInput>,
+) {
+  const hasStateInfo = !!validatedInput.data?.usaState
```
Comment (travisbloom-cb) [inline src/actions/actionCreateUserActionViewKeyRaces.ts:124]: and this

File: src/components/app/recentActivityRow/activityAvatar.tsx [hunk 5]
```diff
@@ -17,6 +17,7 @@ export const ACTIVITY_TYPE_TO_ICON_URL: Record<UserActionType, string> = {
   [UserActionType.LIVE_EVENT]: '/activityFeedIcons/event.svg',
   [UserActionType.TWEET_AT_PERSON]: '/activityFeedIcons/pizza-day.svg',
   [UserActionType.RSVP_EVENT]: '/activityFeedIcons/event-rsvp.svg',
+  [UserActionType.VIEW_KEY_RACES]: '/activityFeedIcons/event-rsvp.svg',
```
[the remark this answers is not in the document]
Comment (travisbloom-cb) [reply src/components/app/recentActivityRow/activityAvatar.tsx:20]: We do want to show this in the recentactivity row, it should be something to the effect of "Someone investigated the key races in their district"

File: src/utils/server/externalOptIn/handleExternalUserActionViewKeyRaces.ts [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+  })
+  const user = existingAction?.user as UserWithRelations
+  const localUser = getLocalUserFromUser(user)
+  const analytics = getServerAnalytics({ userId: user.id, localUser })
+  if (!existingAction?.user) {
+    waitUntil(
+      getServerPeopleAnalytics({ userId: user.id, localUser })
+        .setOnce(mapPersistedLocalUserToAnalyticsProperties(localUser.persisted))
+        .flush(),
+    )
```
[the remark this answers is not in the document]
Comment (olavoparno) [reply src/utils/server/externalOptIn/handleExternalUserActionViewKeyRaces.ts]: thanks man

File: src/actions/actionCreateUserActionViewKeyRaces.ts [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+
+  if (existingViewKeyRacesAction) {
+    logger.info('User already has an existing action')
+
+    const shouldUpdateActionWithAddressInfo =
+      existingViewKeyRacesAction.userActionViewKeyRaces?.usaState !== currentUsaState ||
+      existingViewKeyRacesAction.userActionViewKeyRaces?.usCongressionalDistrict !==
+        currentCongressionalDistrict
+
+    const areNewValuesPresent = currentUsaState !== null || currentCongressionalDistrict !== null
```
Comment (travisbloom-cb) [inline src/actions/actionCreateUserActionViewKeyRaces.ts:131]: this logic is wonky. It should only update an existing key race action if the old one had no metadata for state or district and the new one does OR the old one had state level metadata but no district, and the new one is the same state AND has a district

File: src/utils/server/externalOptIn/handleExternalUserActionViewKeyRaces.ts [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+      result: ExternalUserActionViewKeyRacesResult.EXISTING_ACTION,
+      resultOptions: Object.values(ExternalUserActionViewKeyRacesResult),
+      sessionId,
+      userId,
+    }
+  }
+
+  const currentUsaState = userAddress?.address?.administrativeAreaLevel1 ?? usaState
+
+  const maybeCongressionalDistrict = (await maybeGetCongressionalDistrictFromAddress(
```
Comment (travisbloom-cb) [inline src/utils/server/externalOptIn/handleExternalUserActionViewKeyRaces.ts:86]: you can prob consolidate the logic in the action and this fn

File: src/actions/actionCreateUserActionViewKeyRaces.ts (hunk 7 above)
Comment (olavoparno) [reply src/actions/actionCreateUserActionViewKeyRaces.ts:131]: the problem I face was for cases when the user had no address beforehand and entered a /races/state/district, then after updating their address and if the address had no congressional district due to anything or specificity, we would end up with a false congressional district for the address state

File: src/utils/server/externalOptIn/handleExternalUserActionViewKeyRaces.ts (hunk 8 above)
Comment (olavoparno) [reply src/utils/server/externalOptIn/handleExternalUserActionViewKeyRaces.ts:86]: what do you mean? have a single logic to be reused?

**Review by travisbloom-cb — DISMISSED** [review]
Consider my two comments nice to haves that can be ignored if you prefer. Ready when @mhamendes signs off

File: src/components/app/pageLocationRaceSpecific/index.tsx [hunk 9]
```diff
@@ -55,6 +59,18 @@ export function LocationRaceSpecific({
   const stateDisplayName = stateCode && US_STATE_CODE_TO_DISPLAY_NAME_MAP[stateCode]
   const urls = getIntlUrls(locale)
   const { recommended, others } = findRecommendedCandidate(groups)
+
+  useEffect(() => {
+    async function viewKeyRaces() {
```
Comment (mhamendes) [inline src/components/app/pageLocationRaceSpecific/index.tsx:64]: this function seems unnecessary, you could have the action with a void prefix.

File: src/components/app/pageLocationStateSpecific/index.tsx [hunk 10]
```diff
@@ -49,6 +53,16 @@ export function LocationStateSpecific({
     }),
   )
 
+  useEffect(() => {
+    async function viewKeyRaces() {
```
Comment (mhamendes) [inline src/components/app/pageLocationStateSpecific/index.tsx:57]: same here

**Review by mhamendes — DISMISSED** [review]

**Review by mhamendes — APPROVED** [review]

**Pull request merged**